### ISSUE TITLE
Conf file edition

### ISSUE THREAD

**Issue by yangfl** [pacien/tincapp#30]
I would like to see that *.conf can be directly edited from UI. Especially it would be better if network.conf can be edited on the fly.

**Comment by gappleto97** [comment]
Yes. Please. I've been having to open it in termux for the time being, which is fairly inconvenient.

**Comment by pacien** [comment]
Implementing correctly a graphical interface to edit keys-values would require too much work in my opinion. Contributions are welcome.

The app could also open a file manager in the configuration directory from the configuration screen.

It is currently possible to edit the configuration files from other Android applications that can access the content of the SD card. They can be edited with the Amaze file manager and its included text editor or remotely with Termux+sshd, for example.

**Comment by yangfl** [comment]
I think the info page of a connecting network is just a good template - just make it editable. It won't cover all but still good enough (since other files is already copied to SD card and only network.conf need to be created).

Sorry for my zero knowledge on Android development, otherwise I'd like to contribute :(

**Comment by larsks** [comment]
> The app could also open a file manager in the configuration directory from the configuration screen.

Even better, just add an "open in editor" button. It should be possible to let Android offer the correct applications (I've seen this used to good effect by other apps).

I was just going to add this as a feature request, but since this issue was already open this seemed like a better place for it.

**Comment by pacien** [comment]
I wanted to make the links to the configuration and log directories clickable.
Surprisingly and sadly, there is no way to easily launch a file browser or editor using an Intent on Android. Starting with Android N, the system [forbids such Intents](https://developer.android.com/reference/android/os/FileUriExposedException.html).

**Comment by larsks** [comment]
That is surprising! What a bummer.

**Comment by Beetix** [comment]
On my device (Samsung Galaxy Tab A (2016)) the configuration is saved under `/storage/emulated/0/Android/data/org.pacien.tincapp/files` and termux doesn't have access to this directory (even with storage permission on Android 8.1.0). Hence, I have no way of editing the configuration. In my case, files have to be on the external SD card in order to be edited by the applications.

**Comment by pacien** [comment]
@Beetix wrote:
> On my device (Samsung Galaxy Tab A (2016)) the configuration is saved under `/storage/emulated/0/Android/data/org.pacien.tincapp/files` and termux doesn't have access to this directory (even with storage permission on Android 8.1.0). Hence, I have no way of editing the configuration. In my case, files have to be on the external SD card in order to be edited by the applications.

I've just tested accessing the configuration directory using Termux on an emulator running Android 8.1.0.
You indeed need to grant the storage access permission to Termux and restart that app, closing any running session before it can successfully access the directory.

**Comment by Beetix** [comment]
Thank you. I confirm that it works. It's my mistake... I must have been looking under `/storage/emulated/0/data/` and not `/storage/emulated/0/Android/data/`

**Comment by drboone** [comment]
None of this seems to work in Android 13.  USB Mass Storage has been removed.  MTP is refused access.  Granting permissions via ADB or settings makes no difference in lack of access.  Termux has no access.  I seem to be able to create host files for other nodes, but what's the current solution for extracting the auto-generated public key for the phone?

**Comment by drboone** [comment]
Also, it would appear to be impossible to delete a defined network, due to the permissions issues?

**Comment by pacien** [comment]
Quoting Dennis Boone (2023-09-04 22:02:40)
> None of this seems to work in Android 13.
> USB Mass Storage has been removed.

It should still be there, except if disabled by the device administrator
(company device) or the manufacturer.

It has been renamed to "File transfer mode" on some devices.

> MTP is refused access.

That's not normal.
MTP should have both read and write access to the app's directory.
The device manufacturer might have customised/broken that.

> Granting permissions via ADB or settings makes no difference in
> lack of access.

Only ADB in root mode has access to application data.
This is not available on all devices.

> Termux has no access.

Only on rooted devices and with sudo.

> I seem to be able to create
> host files for other nodes, but what's the current solution for
> extracting the auto-generated public key for the phone?

Only USB/MTP without root, or ADB with root access.

> Also, it would appear to be impossible to delete a defined network,
> due to the permissions issues?

The UI only executes tincctl commands to create networks.
The rest is done through USB/MTP/ADB for configuring or deleting.

**Comment by pacien** [comment]
Since release v0.41, there's now an "Open networks configuration directory" link on the configuration screen, which opens this directory in the standard file manager. The configuration files can then be opened and edited from there using [Editor] or [Acode] for example. The access permissions issue should be resolved too.

[Editor]: https://f-droid.org/packages/org.billthefarmer.editor/
[Acode]: https://play.google.com/store/apps/details?id=com.foxdebug.acodefree

Closing this as resolved; please feel free to provide feedbacks!
